### DISCUSSION TITLE
Could you explain what the macros are doing with z-offsets and surface?

### DISCUSSION THREAD

**Discussion by academo** [jschuh/klipper-macros#141] [Q&A] [answered]
Hello!

I've been using klipper-macros for a while and I've been struggling because I'm calibrating my printer z-offset for a good first layer and I am constantly getting

 ```// Z probe offset or endstop position changed. Run ADJUST_SURFACE_OFFSETS to adjust the offset for all saved surfaces by the change differential, or run ADJUST_SURFACE_OFFSETS IGNORE=1 to hide this message without making changes.``` 

Of course I run the commands, but the first run it shows me some "random" offset and when I run it again it shows all 0.

I don't use any surface feature from the macros. I only have one bed with one surface so I am confused of what the macros are doing with offsets.

I realized that the printer was creating bad first layers because I calibrated the z-offset and forgot to run this command.

Could you please explain me a bit what the macros are doing with the offset and why? is this something I can disable and keep the regular klipper functionality I was used to before using the macros? or maybe it has some advantage I fail to understand and maybe I can benefit from it.

Thanks for such a nice job with these macros and sharing them with the community. So much appreciated.

**Comment by jschuh** [comment, marked as answer]
It's mainly for managing multiple different print surfaces. The general pattern is that you do a first calibration, and after that just use babystepping to fine tune the offset for different surfaces. This is particularly helpful if you have to adjust the the endstop or probe Z offset when you have a bunch of surfaces (because the alternative is to redo each one). But, if you use only one surface, its only real value is that it stores the babystepped offset.

I've been thinking about just silently updating the offset, since that may be less confusing than the warning. But my rationale was that the endstop/probe configuration is a rare event, where it's better to be explicit about what's changing.

**Reply by academo** [reply]
I was doing the babysitting. But then didn't see the warning the first time and when I printed it was all off and I was so confused.

>I've been thinking about just silently updating the offset, since that may be less confusing than the warning

This makes more sense. I have a revo system and there's only one nozzle that is no the same height than the others and when I use it I have to re-calibrate.

Perhaps a config option to either disable surfaces or always update the offset instead of notifying would be great.

Nevertheless I understand it now. Thanks for the explanation!

**Reply by danield340** [reply]
Did we ever get a way to just disable this? It causes more trouble than it's worth imo.
